pkg/image/resolve_test.go

```diff
@@ -0,0 +1,369 @@
+	toolsSystemImages := v3.ToolsSystemImages
+	rkeSystemImages := getRKESystemImages()
+
+	testCases := []struct {
+		tName             string
+		iRkeSystemImages  map[string]v3.RKESystemImages
+		iOsType           OSType
+		oShouldContain    []string
+		oShouldNotContain []string
+	}{
+		{
+			tName:            "linux",
+			iRkeSystemImages: linuxInfo.RKESystemImages,
+			iOsType:          Linux,
+			oShouldContain: []string{
+				toolsSystemImages.PipelineSystemImages.Jenkins,
+				toolsSystemImages.PipelineSystemImages.JenkinsJnlp,
+			},
+			oShouldNotContain: []string{
+				rkeSystemImages.WindowsPodInfraContainer,
+			},
+		},
+		{
+			tName:            "windows",
+			iRkeSystemImages: windowsInfo.RKESystemImages,
+			iOsType:          Windows,
+			oShouldContain: []string{
+				rkeSystemImages.WindowsPodInfraContainer,
+				rkeSystemImages.NginxProxy,
+			},
+			oShouldNotContain: []string{
+				toolsSystemImages.PipelineSystemImages.Jenkins,
+				toolsSystemImages.PipelineSystemImages.JenkinsJnlp,
+			},
+		},
+	}
+
+	assert := assertlib.New(t)
+	for _, cs := range testCases {
+		images, err := fetchImagesFromSystem(cs.iRkeSystemImages, cs.iOsType)
+		assert.Nilf(err, "%s, failed to fetch images from system images", cs.tName)
+
+		if len(cs.oShouldContain) > 0 {
+			assert.Subset(images, cs.oShouldContain, cs.tName)
+		}
+		if len(cs.oShouldNotContain) > 0 {
+			for _, nc := range cs.oShouldNotContain {
+				assert.NotContains(images, nc, cs.tName)
+			}
+		}
+	}
+}
+
+func TestAppendImagesFromRequirement(t *testing.T) {
+	testCases := []struct {
+		tName             string
+		iImages           []string
+		iOsType           OSType
+		oShouldContain    []string
+		oShouldNotContain []string
+	}{
+		{
+			tName:          "linux",
+			iOsType:        Linux,
+			oShouldContain: linuxRequiredImagesNotInSystemCharts,
+		},
+		{
+			tName:             "windows",
+			iOsType:           Windows,
+			oShouldNotContain: linuxRequiredImagesNotInSystemCharts,
+		},
+	}
+
+	assert := assertlib.New(t)
+	for _, cs := range testCases {
+		images := appendImagesFromRequirement(cs.iImages, cs.iOsType)
+
+		if len(cs.oShouldContain) > 0 {
+			assert.Subset(images, cs.oShouldContain, cs.tName)
+		}
+		if len(cs.oShouldNotContain) > 0 {
+			for _, nc := range cs.oShouldNotContain {
+				assert.NotContains(images, nc, cs.tName)
+			}
+		}
+	}
+}
+
+func TestNormalizeImages(t *testing.T) {
+	testCases := []struct {
+		tName             string
+		iRawImages        []string
+		oShouldEqual      []string
+		oShouldContain    []string
+		oShouldNotContain []string
+	}{
+		{
+			tName: "for mirror",
+			iRawImages: []string{
+				"prom/prometheus:v2.0.1",
+				"quay.io/coreos/flannel:v1.2.3",
+				"gcr.io/google_containers/kube-pause:v0.0.1",
+				"test.io/test:v0.0.1", // not in mirror list
+			},
+			oShouldContain: []string{
+				"rancher/coreos-flannel:v1.2.3",
+				"test.io/test:v0.0.1",
+			},
+			oShouldNotContain: []string{
```

Comment:
nit, I think `oShouldEqual` would be enough. The `mirror`, `unique` and `sort` should work together and the output the a raw image list should be explicit.

Reply:
Updated